tirocks/src/snapshot.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+};
+
+use tirocks_sys::{rocksdb_Iterator, rocksdb_Snapshot};
+
+use crate::{
+    db::{RawCfHandle, RawDbRef},
+    option::ReadOptions,
+    properties::table::user::SequenceNumber,
+    Iterable, PinSlice, RawDb, RawIterator, Result,
+};
+
+pub struct RawSnapshot<'a> {
+    ptr: *const rocksdb_Snapshot,
+    _life: PhantomData<&'a ()>,
+}
+
+impl<'a> RawSnapshot<'a> {
+    #[inline]
+    pub(crate) unsafe fn from_ptr(ptr: *const rocksdb_Snapshot) -> RawSnapshot<'a> {
+        RawSnapshot {
+            ptr,
+            _life: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn as_ptr(&self) -> *const rocksdb_Snapshot {
+        self.ptr
+    }
+
+    #[inline]
+    pub fn sequence_number(&self) -> SequenceNumber {
+        let mut n = 0;
+        unsafe {
+            tirocks_sys::crocksdb_get_snapshot_sequence_number(self.ptr, &mut n);
+        }
+        n
+    }
+}
+
+pub struct WithSnapOpt<'a> {
```

Comment:
what's the usage of it

Reply:
An RAII to restore `ReadOption` when access using snapshot.